api/spanner/push_run.go

```diff
@@ -49,3 +67,179 @@ func HandlePushRun(ctx context.Context, auth Authenticator, w http.ResponseWrite
+		return
+	}
+
+	report, err := loadRunReport(ctx, run)
+	if err != nil {
+		shared.GetLogger(ctx).Errorf("Spanner push run failed loading run report: %v", err)
+		return
+	}
+
+	sClient, err = api.SpannerConnect(ctx)
+	if err != nil {
+		shared.GetLogger(ctx).Errorf("Spanner push run failed connecting to spanner: %v", err)
+		return
+	}
+
+	n, err := numRowsToUpload(ctx, sClient, id, report)
+	if err != nil {
+		shared.GetLogger(ctx).Errorf("Spanner push run failed calculating number of rows to upload: %v", err)
+		return
+	}
+
+	shared.GetLogger(ctx).Infof("NOT IMPLEMENTED: Spanner push run would now push run if number of missing rows=%d > 0", n)
+}
+
+// loadRun loads shared.TestRun data from datastore, given an integral ID
+// (Datastore key).
+func loadRun(ctx context.Context, client *datastore.Client, id int64) (*shared.TestRun, error) {
+	logger := shared.GetLogger(ctx)
+
+	logger.Infof("Loading TestRun entity with integra key %d", id)
+
+	var run shared.TestRun
+	err := client.Get(ctx, &datastore.Key{
+		Kind: "TestRun",
+		ID:   id,
+	}, &run)
+	if err != nil {
+		logger.Errorf("Failed to load TestRun entity with integral key %d", id)
+		return nil, err
+	}
+	run.ID = id
+	return &run, nil
+}
+
+// loadRunReport loads a metrics.TestResultsReport using the URL specified in
+// run.
+func loadRunReport(ctx context.Context, run *shared.TestRun) (*metrics.TestResultsReport, error) {
+	logger := shared.GetLogger(ctx)
+
+	if run.RawResultsURL == "" {
+		str := fmt.Sprintf("TestRun entity ID=%d has no RawResultsURL", run.ID)
+		logger.Errorf(str)
+		return nil, errors.New(str)
+	}
+
+	logger.Infof("Reading report from %s", run.RawResultsURL)
+
+	resp, err := http.Get(run.RawResultsURL)
+	if err != nil {
+		logger.Warningf("Failed to load raw results from \"%s\" for run ID=%d", run.RawResultsURL, run.ID)
+		return nil, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		str := fmt.Sprintf("Non-OK HTTP status code of %d from \"%s\" for run ID=%d", resp.StatusCode, run.RawResultsURL, run.ID)
+		logger.Warningf(str)
+		return nil, errors.New(str)
+	}
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		logger.Warningf("Failed to read contents of \"%s\" for run ID=%d", run.RawResultsURL, run.ID)
+		return nil, err
+	}
+	var report metrics.TestResultsReport
+	err = json.Unmarshal(data, &report)
+	if err != nil {
+		logger.Warningf("Failed to unmarshal JSON from \"%s\" for run ID=%d", run.RawResultsURL, run.ID)
+		return nil, err
+	}
+	if len(report.Results) == 0 {
+		str := fmt.Sprintf("Empty report from run ID=%d (%s)", run.ID, run.RawResultsURL)
+		logger.Warningf(str)
+		return nil, errors.New(str)
+	}
+
+	logger.Infof("Read report for run ID=%d", run.ID)
+
+	return &report, nil
+}
+
+// countReportResults counts the number of meaningfully distinct test results
+// detailed in report.
+func countReportResults(ctx context.Context, report *metrics.TestResultsReport) int64 {
+	count := int64(0)
```

Comment:
Are we likely to exceed 2M entries?

Reply:
No, but Spanner will give us an `int64` against which to compare, so this keeps the types the same.